internal/testdeployer/remoteweavelet_test.go

```diff
@@ -0,0 +1,620 @@
+}
+
+// VerifyServerCertificate implements the EnvelopeHandler interface.
+func (d *deployer) VerifyServerCertificate(context.Context, *protos.VerifyServerCertificateRequest) (*protos.VerifyServerCertificateReply, error) {
+	d.t.Fatal("unimplemented")
+	return nil, nil
+}
+
+// deploy creates a new test deployer with a single spawned weavelet.
+func deploy(t *testing.T, ctx context.Context) *deployer {
+	return deployWithInfo(t, ctx, &protos.EnvelopeInfo{
+		App:             "remoteweavelet_test.go",
+		DeploymentId:    fmt.Sprint(os.Getpid()),
+		Id:              uuid.New().String(),
+		InternalAddress: "localhost:0",
+	})
+}
+
+// deploy creates a new test deployer with a single spawned weavelet. The
+// deployer relays the provided EnvelopeInfo to the weavelet.
+func deployWithInfo(t *testing.T, ctx context.Context, info *protos.EnvelopeInfo) *deployer {
+	t.Helper()
+	ctx, cancel := context.WithCancel(ctx)
+
+	// Create pipes to the weavelet.
+	toWeaveletReader, toWeaveletWriter, err := os.Pipe()
+	if err != nil {
+		t.Fatal(err)
+	}
+	toEnvelopeReader, toEnvelopeWriter, err := os.Pipe()
+	if err != nil {
+		t.Fatal(err)
+	}
+	// TODO(mwhittaker): Make sure to close these. This is a bit tricky because
+	// if we close a pipe to the weavelet, it will os.Exit.
+
+	// conn.NewEnvelopeConn blocks performing a handshake with the weavelet, so
+	// we have to run it in a separate goroutine.
+	errs := make(chan error)
+	var env *conn.EnvelopeConn
+	go func() {
+		var err error
+		env, err = conn.NewEnvelopeConn(ctx, toEnvelopeReader, toWeaveletWriter, info)
+		errs <- err
+	}()
+
+	// Create the weavelet.
+	wlet, err := weaver.NewRemoteWeavelet(
+		ctx,
+		codegen.Registered(),
+		runtime.Bootstrap{ToWeaveletFile: toWeaveletReader, ToEnvelopeFile: toEnvelopeWriter},
+		weaver.RemoteWeaveletOptions{},
+	)
+	if err != nil {
+		t.Fatalf("NewRemoteWeavelet: %v", err)
+	}
+
+	// Wait for the EnvelopeConn to finish the handshake.
+	if err := <-errs; err != nil {
+		t.Fatalf("NewEnvelopeConn: %v", err)
+	}
+
+	d := &deployer{
+		t:                t,
+		ctx:              ctx,
+		cancel:           cancel,
+		toWeaveletReader: toWeaveletReader,
+		toWeaveletWriter: toWeaveletWriter,
+		toEnvelopeReader: toEnvelopeReader,
+		toEnvelopeWriter: toEnvelopeWriter,
+		env:              env,
+		wlet:             wlet,
+	}
+	t.Cleanup(d.shutdown)
+	return d
+}
+
+// shutdown shuts down a deployer and its weavelet.
+func (d *deployer) shutdown() {
+	d.cancel()
+	// TODO(mwhittaker): Wait for the weavelet to exit.
+}
+
+// testComponents tests that the components spawned by d are working properly.
+func testComponents(d *deployer) {
+	d.t.Helper()
+	const want = 42
+
+	// Test component a.
+	x, err := d.wlet.GetIntf(reflection.Type[a]())
+	if err != nil {
+		d.t.Fatal(err)
+	}
+	got, err := x.(a).A(d.ctx, want)
+	if err != nil {
+		d.t.Fatal(err)
+	}
+	if got != want {
+		d.t.Fatalf("A(%d): got %d, want %d", want, got, want)
+	}
+
+	// Test component b.
```

Comment:
Just for my understanding: are these tests necessary, given that there's a call chain A -> B -> C ?